Extended diagnostics documentation

dgp1130 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Documentation content changes
- [X] angular.io application / infrastructure changes

Refs #42966.

## What is the new behavior?

Adds a new section for extended diagnostics to describe the feature as whole as well as subpages diving into each currently available diagnostic.

## Does this PR introduce a breaking change?

- [X] No

## Other information

Special thanks to @gkalpak for the docs-infra help! Couldn't have done this without you.

gkalpak committed:
fix(docs-infra): track error docs during `serve-and-sync`

Refs #42966.

Look for changes in error docs (i.e. docs under `aio/content/error/`) in
`authors-package`, so that such docs are tracked when running the
`serve-and-sync` script.

gkalpak committed:
fix(docs-infra): track error docs during `serve-and-sync`

Refs #42966.

Look for changes in error docs (i.e. docs under `aio/content/error/`) in
`authors-package`, so that such docs are tracked when running the
`serve-and-sync` script.

aikidave commented:
@TMDavisGoogle Could you please do an editing pass on this content update? Thank you!

TMDavisGoogle commented on aio/content/extended-diagnostics/NG8101.md:
```diff
@@ -0,0 +1,42 @@
+
+As it stands, `([var])` is actually an [event binding](guide/event-binding) with
+the name `[var]`. The developer probably intended this as a two-way binding to a variable named
+`var` and would be suprised to find that value does not update as expected.
```
and would be suprised to find that value does not update as expected.
```suggestion
`var` and would be surprised to find that value does not update as expected.
```
Also, this phrasing seems too informal. Unless it's crucial to user understanding, I would delete this part of the sentence; it will not change the overall meaning of the statement. If this part of the sentence is necessary, then rewrite to reflect that the value does not update as expected, without the projected emotion.

TMDavisGoogle commented on aio/content/extended-diagnostics/NG8102.md:
```diff
@@ -0,0 +1,85 @@
+might reasonably assume that the right side of the nullish coalescing operator is displayed in some
+case, but this is actually not true, which can lead to confusion about the expected output of the
+program.
+
```
the right side of the nullish coalescing operator is displayed in some case, but this is actually not true, which can lead to confusion about the expected output of the program.
```suggestion
Break this sentence up: "...this is actually not true. This can lead to confusion..."
```

dgp1130 replied:
Broke into two sentences.

TMDavisGoogle commented on aio/content/extended-diagnostics/index.md:
```diff
@@ -0,0 +1,81 @@
+There are many coding patterns that are are technically valid to the compiler or runtime, but which
+may have complex nuances or caveats. These patterns may not have the intended effect expected by a
+developer which often leads to bugs. The Angular compiler includes "extended diagnostics" which
+identify many of these patterns to warn developers about the potential issues and enforce common
```
expected by a developer which often leads to bugs.
```suggestion
insert comma after "developer": "...expected by a developer, which often leads to bugs."
```

TMDavisGoogle commented on aio/content/extended-diagnostics/index.md:
```diff
@@ -0,0 +1,81 @@
+
+There are many coding patterns that are are technically valid to the compiler or runtime, but which
+may have complex nuances or caveats. These patterns may not have the intended effect expected by a
+developer which often leads to bugs. The Angular compiler includes "extended diagnostics" which
```
expected by a developer which often leads to bugs.
```suggestion
insert comma after "developer": "...expected by a developer, which often leads to bugs."

TMDavisGoogle commented on aio/content/extended-diagnostics/index.md:
```diff
@@ -0,0 +1,81 @@
+The Angular team intends to add or enable new extended diagnostics in **minor** versions of Angular
+(see [semver](https://docs.npmjs.com/about-semantic-versioning)). This means that upgrading Angular
+may show new warnings in your existing codebase. This enables the team to deliver features more
+quickly and make extended diagnostics more accessible to developers.
```
and make extended diagnostics
```suggestion

Insert "to" before "make extended": "...quickly and to make extended diagnostics more accessible to developers."
```

TMDavisGoogle commented on aio/content/extended-diagnostics/index.md:
```diff
@@ -0,0 +1,81 @@
+quickly and make extended diagnostics more accessible to developers.
+
+However, setting `"defaultCategory": "error"` will promote such warnings to hard errors. This can
+cause a minor version upgrade to introduce compilation errors which may be seen as a semver
```
introduce compilation errors which may be seen as a semver
```suggestion
Insert comma after "errors":  "...cause a minor version upgrade to introduce compilation errors, which may be seen as a semver..."
```

TMDavisGoogle commented on aio/content/extended-diagnostics/index.md:
```diff
@@ -0,0 +1,81 @@
+non-compliant breaking change. Any new diagnostics can be suppressed or demoted to warnings via the
+above [configuration](#configuration), so the impact of a new diagnostic should be minimal to
+projects that treat extended diagnostics as errors by default. Defaulting to error is a very
+powerful tool, just be aware of this semver caveat when deciding if `error` is the right default for
```
powerful tool,
```suggestion
Replace comma with semicolon:  "...powerful tool;" 
```

TMDavisGoogle reviewed (commented):
Light editing (phrasing, punctuation) done.

dgp1130 commented:
Thanks @TMDavisGoogle, applied your suggestions.

gkalpak reviewed (approved):
:tada:

dgp1130 commented:
I realized I forgot to mention that `strictTemplates` is required (and that certain extended diagnostics may have their own requirements). I added `strictTemplates` to the base doc and also added a "Configuration requirements" section for each check to explain what additional configuration it needs to run.

atscott commented on aio/content/errors/index.md:
```diff
@@ -1 +1 @@
-# Errors List
\ No newline at end of file
+# Errors List
```
Should this file be reverted or regenerated? I'm not sure what this is.

dgp1130 replied:
Ah sorry, I think VSCode adding a trailing newline. Reverted.

atscott commented on aio/content/extended-diagnostics/NG8101.md:
```diff
@@ -0,0 +1,47 @@
+@name Invalid Banana-in-Box
```
Comment on the name which may require renaming a bunch of things (so feel free to ignore): The only place on AIO that refers to this uses "banana in _a_ box" (https://angular.io/guide/ajs-quick-reference#ngmodel). I'm guessing part of the current name was to avoid two capital letters in a row with camel casing?

dgp1130 replied:
We've used "Banana in box" for the name. I don't think that was very intentional, but the name `bananaInBox` will become public API with 13.2. We could try to update all the places it's used in the code base, but that's going to be a lot more than that one "Banana in a box" reference, so I'm inclined to leave this as is, and we can clean up that doc later.

atscott commented on aio/content/extended-diagnostics/NG8102.md:
```diff
@@ -0,0 +1,91 @@
+
+## Configuration requirements
+
+[`strictTemplates`](/guide/template-typecheck#strict-mode) must be enabled for any extended
+diagnostic to emit. [`strictNullChecks`](/guide/template-typecheck#strict-null-checks) must also be
```
Is it worth clarifying that enabling strict templates enables all extended diagnostics automatically? i.e., there's no separate thing you need to "turn on" if you have strict templates enabled.

dgp1130 replied:
Reworded this to be more clear that extended diagnostics are enabled when `strictTemplates` is `true`.

dgp1130 committed:
docs: add extended diagnostics documentation

This includes the initial documentation for extended diagnostics with a page for each of the two initial checks. They follow the same general formula, and hopefully in the future they can be properly generated from metadata rather than copy-pasted as they are currently.

aikithoughts reviewed (approved):
Reviewed-for: global-docs-approvers

atscott reviewed (approved):
reviewed-for: size-tracking

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: size-tracking

AndrewKushnir commented:
This PR was merged into the repository by commit 33f23abe1d79c4309cca8dd852d1ebf82bc336a9.
